lib/CL/devices/tbb/tbb_scheduler.cc

```diff
@@ -0,0 +1,435 @@
+          POCL_MSG_WARN ("TBB: Malformed string in POCL_TBB_PARTITIONER env var: %s\n", ptr);
+        }
+    }
+}
+
+void
+tbb_scheduler_uninit ()
+{
+  unsigned i;
+
+  POCL_FAST_LOCK (scheduler.wq_lock_fast);
+  scheduler.meta_thread_shutdown_requested = 1;
+  pthread_cond_broadcast (&scheduler.wake_meta_thread);
+  POCL_FAST_UNLOCK (scheduler.wq_lock_fast);
+
+  pthread_join (scheduler.meta_thread, NULL);
+
+  POCL_FAST_DESTROY (scheduler.wq_lock_fast);
+  pthread_cond_destroy (&scheduler.wake_meta_thread);
+
+  scheduler.meta_thread_shutdown_requested = 0;
+
+  pocl_aligned_free (scheduler.printf_buf_global_ptr);
+  pocl_aligned_free (scheduler.local_mem_global_ptr);
+}
+
+/* push_command and push_kernel MUST use broadcast and wake up all threads,
+   because commands can be for subdevices (= not all threads) */
+void
+tbb_scheduler_push_command (_cl_command_node *cmd)
+{
+  POCL_FAST_LOCK (scheduler.wq_lock_fast);
+  DL_APPEND (scheduler.work_queue, cmd);
+  pthread_cond_broadcast (&scheduler.wake_meta_thread);
+  POCL_FAST_UNLOCK (scheduler.wq_lock_fast);
+}
+
+/* Internal functions */
+
+class WorkGroupScheduler {
+  kernel_run_command *my_k;
+public:
+  void operator()( const tbb::blocked_range3d<size_t>& r ) const {
+    kernel_run_command *k = my_k;
+    pocl_kernel_metadata_t *meta = k->kernel->meta;
+    const size_t my_thread_id = tbb::this_task_arena::current_thread_index();
+    void *arguments[meta->num_args + meta->num_locals + 1];
+    void *arguments2[meta->num_args + meta->num_locals + 1];
+    char *local_mem = scheduler.local_mem_global_ptr + (scheduler.local_mem_size * my_thread_id);
+    uchar *printf_buffer = scheduler.printf_buf_global_ptr + (scheduler.printf_buf_size * my_thread_id);
+    struct pocl_context pc;
+    /* some random value, doesn't matter as long as it's not a valid bool - to force a first FTZ setup */
+    unsigned current_ftz = 213;
+
+    setup_kernel_arg_array_with_locals ((void **)&arguments, (void **)&arguments2, k, local_mem, scheduler.local_mem_size);
+    memcpy (&pc, &k->pc, sizeof(struct pocl_context));
+
+    // capacity and position already set up
+    pc.printf_buffer = printf_buffer;
+    uint32_t position = 0;
+    pc.printf_buffer_position = &position;
+    assert (pc.printf_buffer != NULL);
+    assert (pc.printf_buffer_capacity > 0);
+    assert (pc.printf_buffer_position != NULL);
+
+    /* Flush to zero is only set once at start of kernel (because FTZ is
+     * a compilation option), but we need to reset rounding mode after every
+     * iteration (since it can be changed during kernel execution). */
+    unsigned flush = k->kernel->program->flush_denorms;
+    if (current_ftz != flush) {
+      pocl_set_ftz (flush);
+      current_ftz = flush;
+    }
+
+    for (size_t x=r.pages().begin(); x!=r.pages().end(); x++) {
+      for (size_t y=r.rows().begin(); y!=r.rows().end(); y++) {
+        for (size_t z=r.cols().begin(); z!=r.cols().end(); z++) {
+          pocl_set_default_rm();
```

Comment:
does it have to set it before each call or is per thread enough?

Reply:
According to the already existing comment, before/after each call.